appliances/Harbor/appliance.sh

```diff
@@ -0,0 +1,358 @@
+    'HARBOR_DB_PASSWORD'        'configure' 'Harbor database password'                          'O|password'
+    'HARBOR_HOSTNAME'           'configure' 'Harbor hostname/IP'                                'O|text'
+    'HARBOR_SSL_CERT'           'configure' 'SSL certificate (.crt)'                            'O|text64'
+    'HARBOR_SSL_KEY'            'configure' 'SSL key (.key)'                                    'O|text64'
+    'HARBOR_PERSISTENT_DEV'     'configure' 'Harbor data persistent device (sda, vda, etc.)'    'O|text'
+)
+
+### Appliance metadata ###############################################
+
+# Appliance metadata
+ONE_SERVICE_NAME='Service Harbor - KVM'
+ONE_SERVICE_VERSION='2.9.4'   #latest
+ONE_SERVICE_BUILD=$(date +%s)
+ONE_SERVICE_SHORT_DESCRIPTION='Appliance running Harbor Docker repository for KVM hosts'
+ONE_SERVICE_DESCRIPTION=$(cat <<EOF
+Appliance with preinstalled Harbor. Run with default values and manually 
+configure it, or use contextualization variables to automate the bootstrap. 
+
+After deploying the appliance, check the status of the deployment in
+/etc/one-appliance/status. You chan check the appliance logs in 
+/var/log/one-appliance/.
+
+In order to configure data persistency, please arrach a secondary disk to the 
+VM, indicate the disk label in HARBOR_PERSISTENT_DEV and launch the appliance.
+
+**WARNING: Do not use localhost or loopback for \`HARBOR_HOSTNAME\`, it
+breaks the service bootstrap. It's necessary to provide a routable name
+or IP. If not provided, it will use the main VM IP.**
+
+
+**WARNING: The appliance does not permit recontextualization. Modifying the
+context variables will not have any real efects on the running instance.**
+EOF
+)
+
+
+
+### Contextualization defaults #######################################
+HARBOR_ADMIN_PASSWORD="${HARBOR_ADMIN_PASSWORD:-Harbor12345}"  # Default Harbor admin password
+HARBOR_DB_PASSWORD="${HARBOR_DB_PASSWORD:-$(gen_db_password)}" # Defaults to autogenerated password
+HARBOR_HOSTNAME="${HARBOR_HOSTNAME:-$(get_local_ip)}" # Listens in all interfaces by default
+HARBOR_SSL_CERT="${HARBOR_SSL_CERT:-AG}" # Defaults to "AG", indicating "AutoGenerated"
+HARBOR_SSL_KEY="${HARBOR_SSL_KEY:-AG}" # Defaults to "AG", indicating "AutoGenerated"
+HARBOR_PERSISTENT_DEV="${HARBOR_PERSISTENT_DEV:-None}" # Listens in all interfaces by default
+
+
+### Installation Stage => Installs requirements, downloads and unpacks Harbor 
+service_install() {
+    msg info "Checking internet access..."
+    check_internet_access
+    install_requirements
+    download_unpack_harbor
+    create_one_service_metadata
+    msg info "Installation phase finished"
+}
+
+### Configuration Stage => Configures Harbor YAML and SSL certificates
+service_configure() { 
+    msg info "Starting configuration..."
+    mount_persistent
+
+    ## SSL CERTS
+
+    if [ ! -d "/root/certs" ]; then
```

Comment:
This check is not really needed, `mkdir -p` is idempotent. You can also use `install -d` as you can specify access rights, owner, group etc. 🤔